File: 0563_Binary_Tree_Tilt/solution.java
package cn.huang.leetcode;

import cn.huang.leetcode.common.TreeNode;

/*
    563. Binary Tree Tilt
 */
public class LC_0563_BinaryTreeTilt {
    int sum = 0;

    public int findTilt(TreeNode root) {
        if(root == null)
            return 0;
        helper(root);

        return sum;
    }

    private int helper(TreeNode root){
        if(root == null)
            return 0;

        int left = helper(root.left);
        int right = helper(root.right);
        this.sum += Math.abs(left - right);
        return root.val + left + right;
    }

}

File: 0563_Binary_Tree_Tilt/solution.cpp
/*
563. Binary Tree Tilt

https://leetcode.com/problems/binary-tree-tilt/

返回树的总tilt值。对于一个节点的tilt值定义为左右子树元素和的绝对之差。
 */
#include <algorithm>

#include "common/TreeNode.h"

using namespace std;

class Solution
{
public:
  int findTilt(TreeNode *root)
  {
    return findTiltSum(root).first;
  }

private:
  pair<int, int> findTiltSum(TreeNode *root)
  {
    if (root == nullptr)
      return {0, 0};
    auto l = findTiltSum(root->left);
    auto r = findTiltSum(root->right);
    return {l.first + r.first + abs(l.second - r.second),
            root->val + l.second + r.second};
  }
};
